### PR TITLE
Genia tokenizer

### PR DESCRIPTION
**Pull request by DeNeutoy**
#### Tokenising 10k pubmed abstracts:
Speed test: genia_tokenizer
iteration 0
iteration 1
iteration 2
Mean (3 runs) 28.98577332496643, Std: 0.9618087746695202.

Speed test: custom_tokenizer
iteration 0
iteration 1
iteration 2
Mean (3 runs) 104.88270195325215, Std: 0.5736707635762457.

Speed test: spacy
iteration 0
iteration 1
iteration 2
Mean (3 runs) 64.94790593783061, Std: 0.6912257483692967.

### CODE REVIEW

**Comment by DeNeutoy** [comment]
I wanted to get several eyes on this because it's complicated (perhaps unnecessarily). However, it is very fast - 2x faster than the Spacy tokenizer which is written in Cython. I'm pretty pleased with that!

**Comment by brendan-ai2** [comment]
Could you elaborate on how the speed tests were performed? If it was just running via `pytest` by hand, I'd be worried about startup costs dominating.

**Comment by danielkingai2** [comment]
I think it would be good to at least know how many of the original spacy tests this tokenizer fails. I imagine that the spacy tokenizer's long set of special strings/patterns that it handles are what makes it slower, although not sure about that.

File: scispacy/genia_tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+# (IL)-mediated", and "p65(RelA)/p50", treat following dashes and
+# slashes as space.  Nested brackets are resolved inside-out;
+# to get this right, add a heuristic considering boundary
+# brackets as "space".
+
+# (First a special case (rareish): "protect" cases with dashes after
+# paranthesized expressions that cannot be abbreviations to avoid
+# breaking up e.g. "(+)-pentazocine". Here, "cannot be abbreviations"
+# is taken as "contains no uppercase charater".)
+_INITIAL.append((re.compile(r'\(([^ ()\[\]{}]+)\)-'), r'-LRB-\1-RRB--'))
```
Comment (danielkingai2) [inline scispacy/genia_tokenizer.py]: I think the comment doesn't match the code here anymore. There is nothing in this regex that excludes uppercase characters.

**Comment by DeNeutoy** [comment]
Hmm I think this fails too many of the spacy tests to be useful. I should have done that earlier...

**Comment by danielkingai2** [comment]
Yeah, sorry it didn't occur to me to suggest that right away

**Pull request closed without merging**